docs/source/site-specific-config.rst

```diff
@@ -0,0 +1,142 @@
+.. _site_specific_config:
+
+Site-Specific Configuration
+***************************
+A site might have compilers that Fab doesn't know about, or prefers
+a different compiler from the Fab default. Fab abstracts the compilers
+and other programs required during building as an instance of a
+:class:`~fab.tools.tool.Tool` class. All tools that Fab knows about, are
+available in a :class:`~fab.tools.tool_repository.ToolRepository`.
+That will include tools that might not be available on the current system.
+
+Each tool belongs to a certain category of
+:class:`~fab.tool.categories.Categories`. A `ToolRepository` can store
+several instances of the same category.
+
+At build time, the user has to create an instance of
+:class:`~fab.tools.tool_box.ToolBox` and pass
+it to the :class:`~fab.build_config.BuildConfig` object. This toolbox
+contains all the tools that will be used during the build process, but
+it can only store one tool per category. If a certain tool should not
+be defined in the toolbox, the default from the `ToolRepository` will
+be used. This is useful for many standard tools like `git`, `rsync`
+etc that de-facto will never be changed.
+
+
+Categories
+==========
+All possible categories are defined in
+:class:`~fab.tool.categories.Categories`. If additional categories
+should be required, they can be added.
+
+Tool
+====
+Each tool must be derived from :class:`~fab.tools.tool.Tool`.
+The base class provides a `run` method, which any tool can
+use to execute a command in a shell. Typically, a tool will
+provide one (or several) custom commands to be used by the steps.
+For example, a compiler instance provides a
+:func:`~fab.tools.compiler.Compiler.compile_file` method.
+This makes sure that no tool-specific command line options need
+to be used in any Fab step, which will allow to replace any tool
+with a different one.
+
+New tools can easily be created, look at
+:class:`~fab.tools.compiler.Gcc` or
+:class:`~fab.tools.compiler.Icc`. Typically, they can just be
+created by providing a different set of parameters in the
+constructor.
+
+This also allows to easily define compiler wrapper. For example,
+if you want to use `mpif90` as compiler, which is a MPI-specific
+wrapper for `ifort`, you can create this class as follows:
```

Comment:
Well, the difficulty is that even though the wrapper is always called `mpif90` (or whatever), but it will need different options to e.g. specify module output path, openmp etc.

But I will need to think about the linker handling, e.g. supporting non-MPI builds?? Do we need that? Obviously, we can use any MPI compiler wrapper for non-MPI programs without much harm (minor link time increase since it might search the MPI libraries??)

Reply:
Done.